### PR TITLE
feat: accept javascript programming language

### LINKED ISSUE #439: Accept Javascript programming language
Javascript is 3rd most used programming language in Indonesia. If we want to promote this for professional we need to allow Javascript as one programming language, either as stdin -> stdout contest, or as function -> output contest

### PR DESCRIPTION
**Pull request by kadeksuryam**
Resolve #439 

For now, I choose to use Node.js version v18 as the Grader Runtime.

Local Testing:
image

Example submission file for the A+A problem:
```
"use strict";
 
process.stdin.resume();
process.stdin.setEncoding("utf-8");
 
let inputString = "";
let currentLine = 0;
 
process.stdin.on("data", (inputStdin) => {
    inputString += inputStdin;
});
 
process.stdin.on("end", (_) => {
    inputString = inputString
        .trim()
        .split("\n")
        .map((string) => {
            return string.trim();
        });
 
    main();
});
 
function readline() {
    return inputString[currentLine++];
}
 
function print(data) {
    console.log(data);
}
 
function main() {
    let n = Number(readline());

    print(n*2);
}
```

### CODE REVIEW

File: judgels-backends/judgels-grader-engines/src/main/java/judgels/gabriel/languages/javascript/JavascriptNode18GradingLanguage.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
+package judgels.gabriel.languages.javascript;
+
+import com.google.common.collect.ImmutableList;
+import java.util.List;
+import judgels.gabriel.api.GradingLanguage;
+
+public class JavascriptNode18GradingLanguage implements GradingLanguage {
```
Comment (fushar) [inline judgels-backends/judgels-grader-engines/src/main/java/judgels/gabriel/languages/javascript/JavascriptNode18GradingLanguage.java]: Nit: The language name is `JavaScript`, not `Javascript`.

Can we reflect this to all filenames and strings?

File: judgels-backends/judgels-grader-engines/src/main/java/judgels/gabriel/languages/javascript/JavascriptNode18GradingLanguage.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    @Override
+    public String getName() {
+        return "Javascript (Node.js 18)";
```
Comment (fushar) [inline judgels-backends/judgels-grader-engines/src/main/java/judgels/gabriel/languages/javascript/JavascriptNode18GradingLanguage.java]: Nit: I'd prefer just `JavaScript (Node 18)`

File: judgels-client/src/modules/api/gabriel/language.js [hunk 3]
```diff
@@ -14,6 +14,7 @@ export const gradingLanguageNamesMap = {
   Python3: 'Python 3',
   PyPy3: 'PyPy 3',
   Rust2021: 'Rust 2021',
+  JavascriptNode18: 'Javascript (Node.js 18)',
```
Comment (fushar) [inline judgels-client/src/modules/api/gabriel/language.js]: Nit: the languages here (and other occurrences in this file) are sorted alphabetically, can you also do that for the new JavaScript language?

File: judgels-backends/judgels-grader-app/Dockerfile [hunk 4]
```diff
@@ -40,6 +42,9 @@ RUN apt-get update \
     && ln -s /usr/bin/gcc-11 /usr/bin/gcc \
     && ln -s /usr/bin/g++-11 /usr/bin/g++ \
     && curl -sSf https://sh.rustup.rs/ | bash -s -- -y -q --default-toolchain=1.57.0 \
+    && curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.39.5/install.sh | bash \
```
Comment (fushar) [inline judgels-backends/judgels-grader-app/Dockerfile]: I think it is safer to install Node from the "official" Ubuntu way, not via third-party nvm

https://www.linode.com/docs/guides/install-nodejs-on-ubuntu-22-04/

What do you think?

**Review by fushar — COMMENTED** [review]
Thanks for the contribution! Really appreciated it.

However, we currently are punting the initiatives of adding language support for "non-CP" languages (JS, TS, PHP). This is because for it to be effective, we need to support "function-style" submission. By this I mean contestants can just fill a skeleton and submit it:

```
function add(a, b) {
  return 0; // complete here
}
```

without actually reading the inputs from stdin, which is not really convenient and intuitive.

The current infrastructure still does not support this.

We may revisit this in the future, but I'll keep this PR open. 😄

File: judgels-backends/judgels-grader-app/Dockerfile (hunk 4 above)
Comment (kadeksuryam) [reply judgels-backends/judgels-grader-app/Dockerfile]: Yes, I will change it. However, the setup that the link you provided, according to https://github.com/nodesource/distributions, the installation scripts 'setup_XX.x' are no longer supported. So, I will follow the official guide instead(adding keyrings, etc)

**Comment by kadeksuryam** [comment]
> Thanks for the contribution! Really appreciated it.
> 
> However, we currently are punting the initiatives of adding language support for "non-CP" languages (JS, TS, PHP). This is because for it to be effective, we need to support "function-style" submission. By this I mean contestants can just fill a skeleton and submit it:
> 
> ```
> function add(a, b) {
>   return 0; // complete here
> }
> ```
> 
> without actually reading the inputs from stdin, which is not really convenient and intuitive.
> 
> The current infrastructure still does not support this.
> 
> We may revisit this in the future, but I'll keep this PR open. 😄

Hi @fushar , when you say 'keep this PR open,' do you mean to keep it open until the infrastructure supports functional-style submission or until I have resolved the previous comments?

Btw, platforms like Codeforces also support Node.js (https://codeforces.com/contest/1881/submission/229061225). Maybe keeps both functional-style and input-output submissions? This PR is for handling the input-output submission. Wdyt?

File: judgels-backends/judgels-grader-app/Dockerfile [hunk 5]
```diff
@@ -19,9 +19,16 @@ ENV RUSTUP_HOME /usr
 ENV XDG_CACHE_HOME /tmp
 ENV JAVA_HOME /usr/lib/jvm/java-11-openjdk-amd64
 ENV PATH $JAVA_HOME/bin:$PATH
+ENV NODE_MAJOR 18
```
Comment (fushar) [inline judgels-backends/judgels-grader-app/Dockerfile]: Nit: this env var is not used for Node / actual grading; it's just for this Dockerfile. I'd suggest just inlining the version (`/node_18.x no distro main ...`) like the other compiler versions here

File: judgels-backends/judgels-grader-app/Dockerfile [hunk 6]
```diff
@@ -19,9 +19,16 @@ ENV RUSTUP_HOME /usr
...
 
 RUN apt-get update \
-    && apt-get -y --no-install-recommends install software-properties-common gpg-agent \
+    && apt-get install -y ca-certificates curl gnupg \
```
Comment (fushar) [inline judgels-backends/judgels-grader-app/Dockerfile]: 1. Add `--no-install-recommends`
2. Remove the second `curl` on line 35 below

File: judgels-backends/judgels-grader-app/Dockerfile [hunk 7]
```diff
@@ -37,6 +44,7 @@ RUN apt-get update \
        openjdk-11-jre \
        pypy3 \
        python3 \
+       nodejs \
```
Comment (fushar) [inline judgels-backends/judgels-grader-app/Dockerfile]: Nit: sort (so put this before `openjdk`)

File: judgels-backends/judgels-grader-engines/src/main/java/judgels/gabriel/languages/GradingLanguageRegistry.java [hunk 8]
```diff
@@ -33,6 +34,7 @@ public class GradingLanguageRegistry {
             new PyPy3GradingLanguage(),
             new Python3GradingLanguage(),
             new Rust2021GradingLanguage(),
+            new JavaScriptNode18GradingLanguage(),
```
Comment (fushar) [inline judgels-backends/judgels-grader-engines/src/main/java/judgels/gabriel/languages/GradingLanguageRegistry.java]: Nit: Sort :)

**Review by fushar — COMMENTED** [review]
> Hi @fushar , when you say 'keep this PR open,' do you mean to keep it open until the infrastructure supports functional-style submission or until I have resolved the previous comments?

Until the infra supports.

> Btw, platforms like Codeforces also support Node.js (https://codeforces.com/contest/1881/submission/229061225). Maybe keeps both functional-style and input-output submissions? This PR is for handling the input-output submission. Wdyt?

I'm sorry, but I'd delay it until the infra supports / there is use case. Reasons:

1. JavaScript is not still popular for CP,
2. we are increasing the Docker image size of the grader without justified reason.

🙏

**Comment by kadeksuryam** [comment]
Okay, I will reopen this PR, but on a different branch. I accidentally used my master branch as the base branch, so I cannot contribute next time until this PR is merged.

**Pull request closed without merging**